### ISSUE TITLE
InterfaceCOLMAP fails on two or more identical cameras

### ISSUE THREAD

**Issue by mteg** [cdcseacave/openMVS#420]
Sometimes InterfaceCOLMAP fails miserably with the following message:

```
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
```

I found out that it happens when two cameras end up sharing the same parameters, such as these cameras 5 and 24:

```
# Camera list with one line of data per camera:
#   CAMERA_ID, MODEL, WIDTH, HEIGHT, PARAMS[]
# Number of cameras: 813
1 PINHOLE 1929 1443 1411.21 1411.21 964.5 721.5
(...)
5 PINHOLE 1927 1442 1408.59 1408.59 963.5 721
(...)
24 PINHOLE 1927 1442 1408.59 1408.59 963.5 721
```

The problem is related to the nature of `std::unordered_set`, which is used for reading the camera list in InterfaceCOLMAP.cpp/ImportScene(). Since the type only holds unique elements, and since uniqueness is defined (through `CameraEqualTo`) as equality of parameters (and not equality of IDs), the camera with ID=24 is never processed in the following loop and thus there is no entry for this camera in `mapCameras[]`:

```
		for (const COLMAP::Camera& colmapCamera: setCameras) {
			mapCameras[colmapCamera.ID] = (uint32_t)scene.platforms.size();

```

Consequently, when the image using this camera is processed in the next loop, `map.Cameras.at()` can't find the camera in the map and thus the crash.
 
```
			Eigen::Map<EVec3d>(&pose.C.x) = -(it.first.q.inverse() * it.first.t);
			Interface::Image image;
			image.name = OPT::strImageFolder+it.first.name;
			image.platformID = mapCameras.at(it.first.idCamera);
```

I added `&& _Left.ID == _Right.ID` to the comparison in `CameraEqualTo` and it prevents the crash, but it seems a dirty fix. Perhaps the whole comparison condition could be replaced by a comparison of IDs. Unfortunately I feel not enough versed in C++ and the codebase to be sure it would not break something else.

PS. Thank you @cdcseacave so much for developing and releasing all this code!

**Comment by cdcseacave** [comment]
pls share a debuging dataset/project

**Comment by LeonChen66** [comment]
I encounter this problem too ...

**Comment by cdcseacave** [comment]
closing as no debug data provided

**Comment by mokrueger** [comment]
@cdcseacave I've encountered the same problem. It is indeed fixable with making the cameras in the cameras.txt unique.
I could provide some sample data if you like.

I am guessing duplicated cameras don't happen in the standard colmap workflow however I've created a custom sparse model (with multiple identical cameras) which resulted in this error (after point_triangulation and image_distortion, trying to call InterfaceCOLMAP). So it seems colmap can handle multiple identical cameras and the Interface cannot. 

Just letting you know that issue exists ;)
Kind regards.
